### ISSUE TITLE
[[Func/training_2/SGD/gradients/gradients/batch_normalization_109/cond_grad/StatelessIf/else/_10305/input/_27305/_3781]]   (1) Failed precondition: Could not find variable batch_normalization_100/gamma. This could mean that the variable has been deleted. In TF1, it can also mean the variable is uninitialized. Debug info: container=localhost, status=Not found: Resource localhost/batch_normalization_100/gamma/N10tensorflow3VarE does not exist. 	 [[{{node ReadVariableOp}}]] 0 successful operations. 0 derived errors ignored

### ISSUE THREAD

**Issue by hebrotem** [tensorflow/tensorflow#58141]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Source

source

### Tensorflow Version

tf 2.10

### Custom Code

Yes

### OS Platform and Distribution

Ubuntu 20.04

### Python version

3.8

### GPU model and memory

RTX2080

### Current Behaviour?

```shell
I expect the model training to begins
```

### Standalone code to reproduce the issue

```shell
__________________________________________________________________________________________________
 Layer (type)                   Output Shape         Param #     Connected to                     
==================================================================================================
 input_1 (InputLayer)           [(None, 224, 224, 3  0           []                               
                                )]                                                                
                                                                                                  
 conv2d (Conv2D)                (None, 112, 112, 64  9472        ['input_1[0][0]']                
                                )                                                                 
                                                                                                  
 batch_normalization (BatchNorm  (None, 112, 112, 64  256        ['conv2d[0][0]']                 
 alization)                     )                                                                 
                                                                                                  
 activation (Activation)        (None, 112, 112, 64  0           ['batch_normalization[0][0]']    
                                )                                                                 
                                                                                                  
 max_pooling2d (MaxPooling2D)   (None, 56, 56, 64)   0           ['activation[0][0]']             
                                                                                                  
 batch_normalization_1 (BatchNo  (None, 56, 56, 64)  256         ['max_pooling2d[0][0]']          
 rmalization)                                                                                     
                                                                                                  
 activation_1 (Activation)      (None, 56, 56, 64)   0           ['batch_normalization_1[0][0]']  
                                                                                                  
 conv2d_1 (Conv2D)              (None, 56, 56, 64)   4160        ['activation_1[0][0]']           
                                                                                                  
...
Total params: 47,705,989
Trainable params: 47,644,549
Non-trainable params: 61,440
```

### Relevant log output

_No response_</details>

**Comment by tilakrayal** [comment]
@hebrotem,
To expedite the trouble-shooting process, could you please provide a complete code you are trying to execute. Thank you!

**Comment by hebrotem** [comment]
model/block.py
**********************
from tensorflow.keras.layers import BatchNormalization
from tensorflow.keras.layers import Conv2D
from tensorflow.keras.layers import UpSampling2D
from tensorflow.keras.layers import Activation
from tensorflow.keras.layers import MaxPool2D
from tensorflow.keras.layers import Add
from tensorflow.keras.layers import Multiply
from tensorflow.keras.layers import Lambda


def residual_block(input, input_channels=None, output_channels=None, kernel_size=(3, 3), stride=1):
    
    if output_channels is None:
        output_channels = input.get_shape()[-1]
    if input_channels is None:
        input_channels = output_channels // 4

    strides = (stride, stride)

    x = BatchNormalization()(input)
    x = Activation('relu')(x)
    x = Conv2D(input_channels, (1, 1))(x)

    x = BatchNormalization()(x)
    x = Activation('relu')(x)
    x = Conv2D(input_channels, kernel_size, padding='same', strides=stride)(x)

    x = BatchNormalization()(x)
    x = Activation('relu')(x)
    x = Conv2D(output_channels, (1, 1), padding='same')(x)

    if input_channels != output_channels or stride != 1:
        input = Conv2D(output_channels, (1, 1), padding='same', strides=strides)(input)

    x = Add()([x, input])
    return x


def attention_block(input, input_channels=None, output_channels=None, encoder_depth=1):
  

    p = 1
    t = 2
    r = 1

    if input_channels is None:
        input_channels = input.shape[-1]
    if output_channels is None:
        output_channels = input_channels

    # First Residual Block
    for i in range(p):
        input = residual_block(input)

    # Trunc Branch
    output_trunk = input
    for i in range(t):
        output_trunk = residual_block(output_trunk)

    # Soft Mask Branch

    ## encoder
    ### first down sampling
    output_soft_mask = MaxPool2D(padding='same')(input)  # 32x32
    for i in range(r):
        output_soft_mask = residual_block(output_soft_mask)

    skip_connections = []
    for i in range(encoder_depth - 1):

        ## skip connections
        output_skip_connection = residual_block(output_soft_mask)
        skip_connections.append(output_skip_connection)
        # print ('skip shape:', output_skip_connection.get_shape())

        ## down sampling
        output_soft_mask = MaxPool2D(padding='same')(output_soft_mask)
        for _ in range(r):
            output_soft_mask = residual_block(output_soft_mask)

            ## decoder
    skip_connections = list(reversed(skip_connections))
    for i in range(encoder_depth - 1):
        ## upsampling
        for _ in range(r):
            output_soft_mask = residual_block(output_soft_mask)
        output_soft_mask = UpSampling2D()(output_soft_mask)
        ## skip connections
        output_soft_mask = Add()([output_soft_mask, skip_connections[i]])

    ### last upsampling
    for i in range(r):
        output_soft_mask = residual_block(output_soft_mask)
    output_soft_mask = UpSampling2D()(output_soft_mask)

    ## Output
    output_soft_mask = Conv2D(input_channels, (1, 1))(output_soft_mask)
    output_soft_mask = Conv2D(input_channels, (1, 1))(output_soft_mask)
    output_soft_mask = Activation('sigmoid')(output_soft_mask)

    # Attention: (1 + output_soft_mask) * output_trunk
    output = Lambda(lambda x: x + 1)(output_soft_mask)
    output = Multiply()([output, output_trunk])  #

    # Last Residual Block
    for i in range(p):
        output = residual_block(output)

    return output

model/model.py
**********************

from  tensorflow.keras.layers import Input
from  tensorflow.keras.layers import Conv2D
from  tensorflow.keras.layers import MaxPool2D
from tensorflow.keras.layers import Dense
from  tensorflow.keras.layers import AveragePooling2D
from keras.layers import Flatten
from keras.layers import Activation
from keras.layers import BatchNormalization
from keras.layers import Dropout
from keras.models import Model
from keras.regularizers import l2

from .blocks import residual_block
from .blocks import attention_block

def AttentionResNet56(shape=(224, 224, 3), n_channels=64, n_classes=100,
                      dropout=0, regularization=0.01):

    regularizer = l2(regularization)

    input_ = Input(shape=shape)
    x = Conv2D(n_channels, (7, 7), strides=(2, 2), padding='same')(input_) # 112x112
    x = BatchNormalization()(x)
    x = Activation('relu')(x)
    x = MaxPool2D(pool_size=(3, 3), strides=(2, 2), padding='same')(x)  # 56x56

    x = residual_block(x, output_channels=n_channels * 4)  # 56x56
    x = attention_block(x, encoder_depth=3)  # bottleneck 7x7

    x = residual_block(x, output_channels=n_channels * 8, stride=2)  # 28x28
    x = attention_block(x, encoder_depth=2)  # bottleneck 7x7

    x = residual_block(x, output_channels=n_channels * 16, stride=2)  # 14x14
    x = attention_block(x, encoder_depth=1)  # bottleneck 7x7

    x = residual_block(x, output_channels=n_channels * 32, stride=2)  # 7x7
    x = residual_block(x, output_channels=n_channels * 32)
    x = residual_block(x, output_channels=n_channels * 32)

    pool_size = (x.shape[1], x.shape[2])
    x = AveragePooling2D(pool_size=pool_size, strides=(1, 1))(x)
    x = Flatten()(x)
    if dropout:
        x = Dropout(dropout)(x)
    output = Dense(n_classes, kernel_regularizer=regularizer, activation='softmax')(x)

    model = Model(input_, output)
    return model

main.py
**********
from tensorflow.keras.metrics import top_k_categorical_accuracy
from models.models import AttentionResNet56

model = AttentionResNet56(shape = (224, 224, 3), n_classes=5)

def top_2_accuracy(in_gt, in_pred):
    return top_k_categorical_accuracy(in_gt, in_pred, k=2)

model.compile(
    optimizer='rmsprop',
  #optimizer=tf.keras.optimizers.SGD(learning_rate=0.001, momentum=0.9), 
  #loss=tf.keras.losses.CategoricalCrossentropy(from_logits=True, label_smoothing=0.1),
    loss='categorical_crossentropy',
  metrics=['categorical_accuracy', top_2_accuracy])

from tensorflow.keras.callbacks import ModelCheckpoint, LearningRateScheduler, EarlyStopping, ReduceLROnPlateau
weight_path="{}_weights.best.hdf5".format('RAN-DR')

checkpoint = ModelCheckpoint(weight_path, monitor='val_loss', verbose=1, 
                             save_best_only=True, mode='min', save_weights_only = True)

reduceLROnPlat = ReduceLROnPlateau(monitor='val_loss', factor=0.8, patience=3, verbose=1, mode='auto', min_delta=0.0001, cooldown=5, min_lr=0.0001)
early = EarlyStopping(monitor="val_loss", 
                      mode="min", 
                      patience=6) # probably needs to be more patient, but kaggle time is limited
callbacks_list = [checkpoint, early, reduceLROnPlat]

model.fit(train_gen, 
                           steps_per_epoch = train_df.shape[0]//batch_size,
                           validation_data = valid_gen, 
                           validation_steps = valid_df.shape[0]//batch_size,
                              epochs = 25, 
                              callbacks = callbacks_list,
                             workers = 0, # tf-generators are not thread-safe
                             use_multiprocessing=False, 
                             max_queue_size = 0, class_weight= class_weights
                             )

**Comment by hebrotem** [comment]
> model/block.py
> 
> from tensorflow.keras.layers import BatchNormalization from tensorflow.keras.layers import Conv2D from tensorflow.keras.layers import UpSampling2D from tensorflow.keras.layers import Activation from tensorflow.keras.layers import MaxPool2D from tensorflow.keras.layers import Add from tensorflow.keras.layers import Multiply from tensorflow.keras.layers import Lambda
> 
> def residual_block(input, input_channels=None, output_channels=None, kernel_size=(3, 3), stride=1):
> 
> ```
> if output_channels is None:
>     output_channels = input.get_shape()[-1]
> if input_channels is None:
>     input_channels = output_channels // 4
> 
> strides = (stride, stride)
> 
> x = BatchNormalization()(input)
> x = Activation('relu')(x)
> x = Conv2D(input_channels, (1, 1))(x)
> 
> x = BatchNormalization()(x)
> x = Activation('relu')(x)
> x = Conv2D(input_channels, kernel_size, padding='same', strides=stride)(x)
> 
> x = BatchNormalization()(x)
> x = Activation('relu')(x)
> x = Conv2D(output_channels, (1, 1), padding='same')(x)
> 
> if input_channels != output_channels or stride != 1:
>     input = Conv2D(output_channels, (1, 1), padding='same', strides=strides)(input)
> 
> x = Add()([x, input])
> return x
> ```
> 
> def attention_block(input, input_channels=None, output_channels=None, encoder_depth=1):
> 
> ```
> p = 1
> t = 2
> r = 1
> 
> if input_channels is None:
>     input_channels = input.shape[-1]
> if output_channels is None:
>     output_channels = input_channels
> 
> # First Residual Block
> for i in range(p):
>     input = residual_block(input)
> 
> # Trunc Branch
> output_trunk = input
> for i in range(t):
>     output_trunk = residual_block(output_trunk)
> 
> # Soft Mask Branch
> 
> ## encoder
> ### first down sampling
> output_soft_mask = MaxPool2D(padding='same')(input)  # 32x32
> for i in range(r):
>     output_soft_mask = residual_block(output_soft_mask)
> 
> skip_connections = []
> for i in range(encoder_depth - 1):
> 
>     ## skip connections
>     output_skip_connection = residual_block(output_soft_mask)
>     skip_connections.append(output_skip_connection)
>     # print ('skip shape:', output_skip_connection.get_shape())
> 
>     ## down sampling
>     output_soft_mask = MaxPool2D(padding='same')(output_soft_mask)
>     for _ in range(r):
>         output_soft_mask = residual_block(output_soft_mask)
> 
>         ## decoder
> skip_connections = list(reversed(skip_connections))
> for i in range(encoder_depth - 1):
>     ## upsampling
>     for _ in range(r):
>         output_soft_mask = residual_block(output_soft_mask)
>     output_soft_mask = UpSampling2D()(output_soft_mask)
>     ## skip connections
>     output_soft_mask = Add()([output_soft_mask, skip_connections[i]])
> 
> ### last upsampling
> for i in range(r):
>     output_soft_mask = residual_block(output_soft_mask)
> output_soft_mask = UpSampling2D()(output_soft_mask)
> 
> ## Output
> output_soft_mask = Conv2D(input_channels, (1, 1))(output_soft_mask)
> output_soft_mask = Conv2D(input_channels, (1, 1))(output_soft_mask)
> output_soft_mask = Activation('sigmoid')(output_soft_mask)
> 
> # Attention: (1 + output_soft_mask) * output_trunk
> output = Lambda(lambda x: x + 1)(output_soft_mask)
> output = Multiply()([output, output_trunk])  #
> 
> # Last Residual Block
> for i in range(p):
>     output = residual_block(output)
> 
> return output
> ```
> 
> model/model.py
> 
> from tensorflow.keras.layers import Input from tensorflow.keras.layers import Conv2D from tensorflow.keras.layers import MaxPool2D from tensorflow.keras.layers import Dense from tensorflow.keras.layers import AveragePooling2D from keras.layers import Flatten from keras.layers import Activation from keras.layers import BatchNormalization from keras.layers import Dropout from keras.models import Model from keras.regularizers import l2
> 
> from .blocks import residual_block from .blocks import attention_block
> 
> def AttentionResNet56(shape=(224, 224, 3), n_channels=64, n_classes=100, dropout=0, regularization=0.01):
> 
> ```
> regularizer = l2(regularization)
> 
> input_ = Input(shape=shape)
> x = Conv2D(n_channels, (7, 7), strides=(2, 2), padding='same')(input_) # 112x112
> x = BatchNormalization()(x)
> x = Activation('relu')(x)
> x = MaxPool2D(pool_size=(3, 3), strides=(2, 2), padding='same')(x)  # 56x56
> 
> x = residual_block(x, output_channels=n_channels * 4)  # 56x56
> x = attention_block(x, encoder_depth=3)  # bottleneck 7x7
> 
> x = residual_block(x, output_channels=n_channels * 8, stride=2)  # 28x28
> x = attention_block(x, encoder_depth=2)  # bottleneck 7x7
> 
> x = residual_block(x, output_channels=n_channels * 16, stride=2)  # 14x14
> x = attention_block(x, encoder_depth=1)  # bottleneck 7x7
> 
> x = residual_block(x, output_channels=n_channels * 32, stride=2)  # 7x7
> x = residual_block(x, output_channels=n_channels * 32)
> x = residual_block(x, output_channels=n_channels * 32)
> 
> pool_size = (x.shape[1], x.shape[2])
> x = AveragePooling2D(pool_size=pool_size, strides=(1, 1))(x)
> x = Flatten()(x)
> if dropout:
>     x = Dropout(dropout)(x)
> output = Dense(n_classes, kernel_regularizer=regularizer, activation='softmax')(x)
> 
> model = Model(input_, output)
> return model
> ```
> 
> main.py
> 
> from tensorflow.keras.metrics import top_k_categorical_accuracy from models.models import AttentionResNet56
> 
> model = AttentionResNet56(shape = (224, 224, 3), n_classes=5)
> 
> def top_2_accuracy(in_gt, in_pred): return top_k_categorical_accuracy(in_gt, in_pred, k=2)
> 
> model.compile( optimizer='rmsprop', #optimizer=tf.keras.optimizers.SGD(learning_rate=0.001, momentum=0.9), #loss=tf.keras.losses.CategoricalCrossentropy(from_logits=True, label_smoothing=0.1), loss='categorical_crossentropy', metrics=['categorical_accuracy', top_2_accuracy])
> 
> from tensorflow.keras.callbacks import ModelCheckpoint, LearningRateScheduler, EarlyStopping, ReduceLROnPlateau weight_path="{}_weights.best.hdf5".format('RAN-DR')
> 
> checkpoint = ModelCheckpoint(weight_path, monitor='val_loss', verbose=1, save_best_only=True, mode='min', save_weights_only = True)
> 
> reduceLROnPlat = ReduceLROnPlateau(monitor='val_loss', factor=0.8, patience=3, verbose=1, mode='auto', min_delta=0.0001, cooldown=5, min_lr=0.0001) early = EarlyStopping(monitor="val_loss", mode="min", patience=6) # probably needs to be more patient, but kaggle time is limited callbacks_list = [checkpoint, early, reduceLROnPlat]
> 
> model.fit(train_gen, steps_per_epoch = train_df.shape[0]//batch_size, validation_data = valid_gen, validation_steps = valid_df.shape[0]//batch_size, epochs = 25, callbacks = callbacks_list, workers = 0, # tf-generators are not thread-safe use_multiprocessing=False, max_queue_size = 0, class_weight= class_weights )

**Comment by tilakrayal** [comment]
@hebrotem,
Code shared is full of [indentation errors](https://colab.research.google.com/gist/tilakrayal/64cbc51b2cb4a7df11cd6a39f76a3f95/untitled708.ipynb), please share a colab gist with issue reported or simple stand alone indented code with all dependencies such that we can replicate the issue reported. Thank you!

**Comment by hebrotem** [comment]
Thank you for your response, @tilakrayal. I think the code was messed up after pasting it here. Also, I am only getting this error when I run the code on my local machine with RTX2080Ti GPU. I have been able to run it without any error on kaggle. If you don't mind I can share you the code directly?

**Comment by hebrotem** [comment]
> @hebrotem, Code shared is full of [indentation errors](https://colab.research.google.com/gist/tilakrayal/64cbc51b2cb4a7df11cd6a39f76a3f95/untitled708.ipynb), please share a colab gist with issue reported or simple stand alone indented code with all dependencies such that we can replicate the issue reported. Thank you!

please check this link for the colab gist https://gist.github.com/hebrotem/2ad4fd6df14261a90128db8e45b4052e

**Comment by tilakrayal** [comment]
@hebrotem,
Thanks for opening this issue. Development of keras moved to another [repository](https://github.com/keras-team/keras/issues). 
 
 
Could you please post this issue on keras-team/keras [repo](https://github.com/keras-team/keras/issues).
To know more please refer:
https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999
Thank you!
